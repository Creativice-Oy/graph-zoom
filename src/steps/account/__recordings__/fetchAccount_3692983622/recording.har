{
  "log": {
    "_recordingName": "fetchAccount",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "4.3.0"
    },
    "entries": [
      {
        "_id": "32b6c540dbe7339d81f7948939897f76",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "api.zoom.us"
            }
          ],
          "headersSize": 862,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://api.zoom.us/v2/users/me"
        },
        "response": {
          "bodySize": 739,
          "content": {
            "mimeType": "application/json;charset=UTF-8",
            "size": 739,
            "text": "{\"id\":\"H9Yx0-O8QUCXVIseYAsjaA\",\"first_name\":\"JupiterOne\",\"last_name\":\"Development\",\"email\":\"development@jupiterone.dev\",\"type\":2,\"role_name\":\"Owner\",\"pmi\":4440405982,\"use_pmi\":false,\"personal_meeting_url\":\"https://zoom.us/j/4440405982?pwd=c0Jidml0ZS9ZRzA3dEZWSTNHNHlPZz09\",\"timezone\":\"America/New_York\",\"verified\":1,\"dept\":\"\",\"created_at\":\"2021-10-27T00:00:49Z\",\"last_login_time\":\"2021-11-02T09:51:57Z\",\"host_key\":\"815201\",\"cms_user_id\":\"\",\"jid\":\"h9yx0-o8qucxviseyasjaa@xmpp.zoom.us\",\"group_ids\":[\"ppUAdwEBRFCdRE0mfLttog\"],\"im_group_ids\":[],\"account_id\":\"N0oChUdLTrCZLfk9JXp3Hw\",\"language\":\"\",\"phone_country\":\"\",\"phone_number\":\"\",\"status\":\"active\",\"job_title\":\"\",\"location\":\"\",\"login_types\":[100],\"role_id\":\"0\",\"account_number\":7011249145}"
          },
          "cookies": [
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "zm_aid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "zm_haid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "zm_tmaid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "zm_htmaid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "httpOnly": true,
              "name": "cred",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "_zm_ctaid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "_zm_chtaid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "date",
              "value": "Wed, 03 Nov 2021 04:43:44 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json;charset=UTF-8"
            },
            {
              "name": "content-length",
              "value": "739"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "x-zm-trackingid",
              "value": "v=2.0;clid=aw1;rid=WEB_6b57aea1eaef35fe8e753d8acb99267a"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "cache-control",
              "value": "no-cache, no-store, must-revalidate, no-transform"
            },
            {
              "name": "pragma",
              "value": "no-cache"
            },
            {
              "name": "expires",
              "value": "Thu, 01 Jan 1970 00:00:00 GMT"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "vary",
              "value": "Origin, Access-Control-Request-Method, Access-Control-Request-Headers"
            },
            {
              "name": "x-ratelimit-category",
              "value": "Light"
            },
            {
              "name": "x-zm-zoneid",
              "value": "VA2"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            },
            {
              "name": "referrer-policy",
              "value": "strict-origin-when-cross-origin"
            }
          ],
          "headersSize": 1347,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-11-03T04:43:43.081Z",
        "time": 1176,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 1176
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
