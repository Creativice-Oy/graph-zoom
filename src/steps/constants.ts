import {
  RelationshipClass,
  StepEntityMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';

export enum IntegrationSteps {
  USERS = 'fetch-users',
  GROUPS = 'fetch-groups',
  BUILD_USER_AND_GROUP_RELATIONSHIP = 'build-user-and-group-relationship',
  ROLES = 'fetch-roles',
}

export const Entities: Record<'USER' | 'GROUP' | 'ROLE', StepEntityMetadata> = {
  USER: {
    resourceName: 'User',
    _type: 'zoom_user',
    _class: 'User',
  },
  GROUP: {
    resourceName: 'Group',
    _type: 'zoom_group',
    _class: 'Group',
  },
  ROLE: {
    resourceName: 'Role',
    _type: 'zoom_role',
    _class: 'AccessRole',
  },
};

export const Relationships: Record<
  'GROUP_HAS_USER',
  StepRelationshipMetadata
> = {
  GROUP_HAS_USER: {
    _type: 'zoom_group_has_user',
    _class: RelationshipClass.HAS,
    sourceType: Entities.GROUP._type,
    targetType: Entities.USER._type,
  },
};
