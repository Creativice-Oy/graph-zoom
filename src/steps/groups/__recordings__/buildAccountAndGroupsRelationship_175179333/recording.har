{
  "log": {
    "_recordingName": "buildAccountAndGroupsRelationship",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "4.3.0"
    },
    "entries": [
      {
        "_id": "32b6c540dbe7339d81f7948939897f76",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "api.zoom.us"
            }
          ],
          "headersSize": 862,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://api.zoom.us/v2/users/me"
        },
        "response": {
          "bodySize": 739,
          "content": {
            "mimeType": "application/json;charset=UTF-8",
            "size": 739,
            "text": "{\"id\":\"H9Yx0-O8QUCXVIseYAsjaA\",\"first_name\":\"JupiterOne\",\"last_name\":\"Development\",\"email\":\"development@jupiterone.dev\",\"type\":2,\"role_name\":\"Owner\",\"pmi\":4440405982,\"use_pmi\":false,\"personal_meeting_url\":\"https://zoom.us/j/4440405982?pwd=c0Jidml0ZS9ZRzA3dEZWSTNHNHlPZz09\",\"timezone\":\"America/New_York\",\"verified\":1,\"dept\":\"\",\"created_at\":\"2021-10-27T00:00:49Z\",\"last_login_time\":\"2021-11-02T09:51:57Z\",\"host_key\":\"815201\",\"cms_user_id\":\"\",\"jid\":\"h9yx0-o8qucxviseyasjaa@xmpp.zoom.us\",\"group_ids\":[\"ppUAdwEBRFCdRE0mfLttog\"],\"im_group_ids\":[],\"account_id\":\"N0oChUdLTrCZLfk9JXp3Hw\",\"language\":\"\",\"phone_country\":\"\",\"phone_number\":\"\",\"status\":\"active\",\"job_title\":\"\",\"location\":\"\",\"login_types\":[100],\"role_id\":\"0\",\"account_number\":7011249145}"
          },
          "cookies": [
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "zm_aid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "zm_haid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "zm_tmaid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "zm_htmaid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "httpOnly": true,
              "name": "cred",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "_zm_ctaid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "_zm_chtaid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "date",
              "value": "Wed, 03 Nov 2021 05:00:24 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json;charset=UTF-8"
            },
            {
              "name": "content-length",
              "value": "739"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "x-zm-trackingid",
              "value": "v=2.0;clid=aw1;rid=WEB_ff81e1283a2aa139c3bb109408d22338"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "cache-control",
              "value": "no-cache, no-store, must-revalidate, no-transform"
            },
            {
              "name": "pragma",
              "value": "no-cache"
            },
            {
              "name": "expires",
              "value": "Thu, 01 Jan 1970 00:00:00 GMT"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "vary",
              "value": "Origin, Access-Control-Request-Method, Access-Control-Request-Headers"
            },
            {
              "name": "x-ratelimit-category",
              "value": "Light"
            },
            {
              "name": "x-zm-zoneid",
              "value": "VA2"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            },
            {
              "name": "referrer-policy",
              "value": "strict-origin-when-cross-origin"
            }
          ],
          "headersSize": 1347,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-11-03T05:00:23.589Z",
        "time": 1730,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 1730
        }
      },
      {
        "_id": "7892a632e9fd0e489b37b35fcbd815a2",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "api.zoom.us"
            }
          ],
          "headersSize": 860,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://api.zoom.us/v2/groups"
        },
        "response": {
          "bodySize": 258,
          "content": {
            "mimeType": "application/json;charset=UTF-8",
            "size": 258,
            "text": "{\"total_records\":2,\"groups\":[{\"id\":\"ppUAdwEBRFCdRE0mfLttog\",\"name\":\"Test Group\",\"total_members\":2,\"in_directory\":false,\"directory_privacy\":2},{\"id\":\"dSHPJbpfQo6uS1a0_iznyg\",\"name\":\"Test Group 2\",\"total_members\":0,\"in_directory\":false,\"directory_privacy\":2}]}"
          },
          "cookies": [
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "zm_aid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "zm_haid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "zm_tmaid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "zm_htmaid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "httpOnly": true,
              "name": "cred",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "_zm_ctaid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".zoom.us",
              "expires": "1970-01-01T00:00:10.000Z",
              "httpOnly": true,
              "name": "_zm_chtaid",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "date",
              "value": "Wed, 03 Nov 2021 05:00:26 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json;charset=UTF-8"
            },
            {
              "name": "content-length",
              "value": "258"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "x-zm-trackingid",
              "value": "v=2.0;clid=aw1;rid=WEB_f0a5fb42877dd0301eede101d30665f1"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "cache-control",
              "value": "no-cache, no-store, must-revalidate, no-transform"
            },
            {
              "name": "pragma",
              "value": "no-cache"
            },
            {
              "name": "expires",
              "value": "Thu, 01 Jan 1970 00:00:00 GMT"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "vary",
              "value": "Origin, Access-Control-Request-Method, Access-Control-Request-Headers"
            },
            {
              "name": "x-ratelimit-category",
              "value": "Medium"
            },
            {
              "name": "x-zm-zoneid",
              "value": "VA2"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            },
            {
              "name": "referrer-policy",
              "value": "strict-origin-when-cross-origin"
            }
          ],
          "headersSize": 1348,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-11-03T05:00:25.328Z",
        "time": 905,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 905
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
