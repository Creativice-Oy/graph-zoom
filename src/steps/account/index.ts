import {
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';

import { IntegrationConfig } from '../../config';
import { createAPIClient } from '../../client';
import { Entities, IntegrationSteps } from '../constants';
import { createAccountEntity } from './converters';

export async function fetchAccount({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  const user = await apiClient.getCurrentUser();

  await jobState.addEntity(createAccountEntity(user));
}

export const accountSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: IntegrationSteps.ACCOUNT,
    name: 'Fetch Account',
    entities: [Entities.ACCOUNT],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchAccount,
  },
];
