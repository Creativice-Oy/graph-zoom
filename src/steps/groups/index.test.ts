import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';
import { IntegrationConfig } from '../../config';
import {
  buildUserAndGroupsRelationship,
  buildAccountAndGroupsRelationship,
  fetchGroups,
} from '.';
import { integrationConfig } from '../../../test/config';
import { setupZoomRecording } from '../../../test/recording';
import { fetchUsers } from '../users';
import { Relationships } from '../constants';
import { fetchAccount } from '../account';

describe('#fetchGroups', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupZoomRecording({
      directory: __dirname,
      name: 'fetchGroups',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await fetchGroups(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      collectedEntities: context.jobState.collectedEntities,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    const groups = context.jobState.collectedEntities.filter((e) =>
      e._type.includes('zoom_group'),
    );
    expect(groups.length).toBeGreaterThan(0);
    expect(groups).toMatchGraphObjectSchema({
      _class: ['Group'],
      schema: {
        additionalProperties: false,
        properties: {
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          _type: { const: 'zoom_group' },
          id: { type: 'string' },
          name: { type: 'string' },
          totalMembers: { type: 'number' },
        },
      },
    });
  });
});

describe('#buildUserAndGroupsRelationship', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupZoomRecording({
      directory: __dirname,
      name: 'buildUserAndGroupsRelationship',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should build group and user relationship', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await fetchUsers(context);
    await fetchGroups(context);
    await buildUserAndGroupsRelationship(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      collectedEntities: context.jobState.collectedEntities,
      encounteredTypes: context.jobState.encounteredTypes,
      collectedRelationships: context.jobState.collectedRelationships,
    }).toMatchSnapshot();

    const users = context.jobState.collectedEntities.filter((e) =>
      e._type.includes('zoom_user'),
    );
    expect(users.length).toBeGreaterThan(0);
    expect(users).toMatchGraphObjectSchema({
      _class: ['User'],
      schema: {
        additionalProperties: false,
        properties: {
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          _type: { const: 'zoom_user' },
          name: { type: 'string' },
          username: { type: 'string' },
          firstName: { type: 'string' },
          lastName: { type: 'string' },
          email: { type: 'string' },
          type: { type: 'number' },
          pmi: { type: 'number' },
          timezone: { type: 'string' },
          verified: { type: 'number' },
          createdAt: { type: 'string' },
          lastLoginTime: { type: 'string' },
          lastClientVersion: { type: 'string' },
          picUrl: { type: 'string' },
          language: { type: 'string' },
          phoneNumber: { type: 'string' },
          status: { type: 'string' },
          roleId: { type: 'string' },
          dept: { type: 'string' },
          groupIds: {
            type: 'array',
            items: { type: 'string' },
          },
          hostKey: { type: 'string' },
        },
      },
    });

    const groups = context.jobState.collectedEntities.filter((e) =>
      e._type.includes('zoom_group'),
    );
    expect(groups.length).toBeGreaterThan(0);
    expect(groups).toMatchGraphObjectSchema({
      _class: ['Group'],
      schema: {
        additionalProperties: false,
        properties: {
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          _type: { const: 'zoom_group' },
          id: { type: 'string' },
          name: { type: 'string' },
          totalMembers: { type: 'number' },
        },
      },
    });

    expect(
      context.jobState.collectedRelationships.filter(
        (e) => e._type === Relationships.GROUP_HAS_USER._type,
      ),
    ).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'HAS' },
          _type: {
            const: 'zoom_group_has_user',
          },
        },
      },
    });
  });
});

describe('#buildAccountAndGroupsRelationship', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupZoomRecording({
      directory: __dirname,
      name: 'buildAccountAndGroupsRelationship',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should build group and account relationship', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await fetchAccount(context);
    await fetchGroups(context);
    await buildAccountAndGroupsRelationship(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      collectedEntities: context.jobState.collectedEntities,
      encounteredTypes: context.jobState.encounteredTypes,
      collectedRelationships: context.jobState.collectedRelationships,
    }).toMatchSnapshot();

    const account = context.jobState.collectedEntities.filter((e) =>
      e._type.includes('zoom_account'),
    );
    expect(account.length).toBe(1);
    expect(account).toMatchGraphObjectSchema({
      _class: ['User'],
      schema: {
        additionalProperties: false,
        properties: {
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          _type: { const: 'zoom_account' },
          name: { type: 'string' },
          username: { type: 'string' },
          email: { type: 'string' },
        },
      },
    });

    const groups = context.jobState.collectedEntities.filter((e) =>
      e._type.includes('zoom_group'),
    );
    expect(groups.length).toBeGreaterThan(0);
    expect(groups).toMatchGraphObjectSchema({
      _class: ['Group'],
      schema: {
        additionalProperties: false,
        properties: {
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          _type: { const: 'zoom_group' },
          id: { type: 'string' },
          name: { type: 'string' },
          totalMembers: { type: 'number' },
        },
      },
    });

    expect(
      context.jobState.collectedRelationships.filter(
        (e) => e._type === Relationships.ACCOUNT_HAS_GROUP._type,
      ),
    ).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'HAS' },
          _type: {
            const: 'zoom_account_has_group',
          },
        },
      },
    });
  });
});
