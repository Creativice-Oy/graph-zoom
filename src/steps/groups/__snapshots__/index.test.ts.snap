// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#buildUserAndGroupsRelationship should build group and user relationship 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "zoom_user:H9Yx0-O8QUCXVIseYAsjaA",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "created_at": "2021-10-27T00:00:49Z",
            "email": "development@jupiterone.dev",
            "first_name": "JupiterOne",
            "group_ids": Array [
              "ppUAdwEBRFCdRE0mfLttog",
            ],
            "id": "H9Yx0-O8QUCXVIseYAsjaA",
            "language": "",
            "last_login_time": "2021-10-28T03:29:30Z",
            "last_name": "Development",
            "phone_number": "",
            "pmi": 4440405982,
            "role_id": "0",
            "status": "active",
            "timezone": "America/New_York",
            "type": 2,
            "verified": 1,
          },
        },
      ],
      "_type": "zoom_user",
      "active": true,
      "createdAt": "2021-10-27T00:00:49Z",
      "createdOn": undefined,
      "displayName": "JupiterOne Development",
      "email": "development@jupiterone.dev",
      "firstName": "JupiterOne",
      "id": "H9Yx0-O8QUCXVIseYAsjaA",
      "language": "",
      "lastClientVersion": undefined,
      "lastLoginTime": "2021-10-28T03:29:30Z",
      "lastName": "Development",
      "name": "JupiterOne Development",
      "phoneNumber": "",
      "picUrl": undefined,
      "pmi": 4440405982,
      "roleId": "0",
      "status": "active",
      "timezone": "America/New_York",
      "type": 2,
      "username": "development@jupiterone.dev",
      "verified": 1,
    },
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "zoom_group:ppUAdwEBRFCdRE0mfLttog",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "directory_privacy": 2,
            "id": "ppUAdwEBRFCdRE0mfLttog",
            "in_directory": false,
            "name": "Test Group",
            "total_members": 1,
          },
        },
      ],
      "_type": "zoom_group",
      "createdOn": undefined,
      "displayName": "Test Group",
      "id": "ppUAdwEBRFCdRE0mfLttog",
      "name": "Test Group",
      "totalMembers": 1,
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "zoom_group:ppUAdwEBRFCdRE0mfLttog",
      "_key": "zoom_group:ppUAdwEBRFCdRE0mfLttog|has|zoom_user:H9Yx0-O8QUCXVIseYAsjaA",
      "_toEntityKey": "zoom_user:H9Yx0-O8QUCXVIseYAsjaA",
      "_type": "zoom_group_has_user",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "zoom_user",
    "zoom_group",
    "zoom_group_has_user",
  ],
  "numCollectedEntities": 2,
}
`;

exports[`#fetchGroups should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "zoom_group:ppUAdwEBRFCdRE0mfLttog",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "directory_privacy": 2,
            "id": "ppUAdwEBRFCdRE0mfLttog",
            "in_directory": false,
            "name": "Test Group",
            "total_members": 1,
          },
        },
      ],
      "_type": "zoom_group",
      "createdOn": undefined,
      "displayName": "Test Group",
      "id": "ppUAdwEBRFCdRE0mfLttog",
      "name": "Test Group",
      "totalMembers": 1,
    },
  ],
  "encounteredTypes": Array [
    "zoom_group",
  ],
  "numCollectedEntities": 1,
}
`;
